File: disjoint_set.cpp
#include <iostream>
#include <vector>

class UnionFind {
public:
    UnionFind(int size) : root(size) {
        for (int i = 0; i < size; i++) {
            root[i] = i;
        }
    }

    int find(int x) {
        if (x == root[x]) {
            return x;
        }
        return find(root[x]);
    }

    void unionSet(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        if (rootX != rootY) {
            root[rootY] = rootX;
        }
    }

private:
    std::vector<int> root;
};

int main() {
    UnionFind uf(10);

    uf.unionSet(1, 2);
    uf.unionSet(2, 3);
    uf.unionSet(4, 5);
    uf.unionSet(6, 7);
    uf.unionSet(7, 8);

    for (int i = 0; i < 10; i++) {
        std::cout << "Αντιπρόσωπος του κόμβου " << i << ": " << uf.find(i) << std::endl;
    }

    return 0;
}

File: disjoint_set.py
class UnionFind:
    def __init__(self, size):
        self.root = [i for i in range(size)]

    def find(self, x):
        if x == self.root[x]:
            return x
        return self.find(self.root[x])

    def union(self, x, y):
        rootX = self.find(x)
        rootY = self.find(y)
        if rootX != rootY:
            self.root[rootY] = rootX

# Δημιουργία ενός αντικειμένου UnionFind με 10 κόμβους
uf = UnionFind(10)

# Εκτέλεση κάποιων λειτουργιών ένωσης
uf.union(1, 2)
uf.union(2, 3)
uf.union(4, 5)
uf.union(6, 7)
uf.union(7, 8)

# Εκτύπωση των αντιπροσώπων κάθε κόμβου
for i in range(10):
    print(f"Αντιπρόσωπος του κόμβου {i}: {uf.find(i)}")
